sw/device/lib/crypto/impl/rsa/rsa_signature.c

```diff
@@ -0,0 +1,374 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/impl/rsa/rsa_signature.h"
+
+#include "sw/device/lib/base/hardened.h"
+#include "sw/device/lib/base/hardened_memory.h"
+#include "sw/device/lib/crypto/impl/rsa/rsa_modexp.h"
+#include "sw/device/lib/crypto/impl/sha2/sha256.h"
+#include "sw/device/lib/crypto/impl/sha2/sha512.h"
+
+// Module ID for status codes.
+#define MODULE_ID MAKE_MODULE_ID('r', 's', 'v')
+
+/**
+ * Digest identifiers for different hash functions (little-endian).
+ *
+ * See Note 1 in RFC 8017.
+ */
+static const uint8_t kSha256DigestIdentifier[] = {
+    0x20, 0x04, 0x00, 0x05, 0x01, 0x02, 0x04, 0x03, 0x65, 0x01,
+    0x48, 0x86, 0x60, 0x09, 0x06, 0x0d, 0x30, 0x31, 0x30,
+};
+static const uint8_t kSha384DigestIdentifier[] = {
+    0x30, 0x04, 0x00, 0x05, 0x02, 0x02, 0x04, 0x03, 0x65, 0x01,
+    0x48, 0x86, 0x60, 0x09, 0x06, 0x0d, 0x30, 0x41, 0x30,
+};
+static const uint8_t kSha512DigestIdentifier[] = {
+    0x40, 0x04, 0x00, 0x05, 0x03, 0x02, 0x04, 0x03, 0x65, 0x01,
+    0x48, 0x86, 0x60, 0x09, 0x06, 0x0d, 0x30, 0x51, 0x30,
+};
+
+/**
+ * Get the length of the DER encoding for the given hash function's digests.
+ *
+ * See RFC 8017, Appendix B.1. The encoding consists of the digest algorithm
+ * identifier and then the digest itself.
+ *
+ * @param hash_mode Hash function to use.
+ * @param[out] len Byte-length of the DER encoding of the digest.
+ * @param OTCRYPTO_BAD_ARGS if the hash function is not valid, otherwise OK.
+ */
+status_t digest_info_length_get(const rsa_signature_hash_t hash_mode,
+                                size_t *len) {
+  switch (hash_mode) {
+    case kRsaSignatureHashSha256:
+      *len = sizeof(kSha256DigestIdentifier) + kSha256DigestBytes;
+      return OTCRYPTO_OK;
+    case kRsaSignatureHashSha384:
+      *len = sizeof(kSha384DigestIdentifier) + kSha384DigestBytes;
+      return OTCRYPTO_OK;
+    case kRsaSignatureHashSha512:
+      *len = sizeof(kSha512DigestIdentifier) + kSha512DigestBytes;
+      return OTCRYPTO_OK;
+    default:
+      // Unrecognized hash function.
+      return OTCRYPTO_BAD_ARGS;
+  };
+
+  // Unreachable.
+  HARDENED_TRAP();
+  return OTCRYPTO_FATAL_ERR;
+}
+
+/**
+ * Get the DER encoding for the hash function's digests.
+ *
+ * See RFC 8017, Appendix B.1.
+ *
+ * The caller must ensure that enough space is allocated for the encoding; use
+ * `digest_info_length()` to check before calling this function.
+ *
+ * The encoding produced is little-endian, reversed compared to the RFC.
+ *
+ * @param message Message to hash.
+ * @param message_len Message length in bytes.
+ * @param hash_mode Hash function to use.
+ * @param[out] encoding DER encoding of the digest.
+ * @param OTCRYPTO_BAD_ARGS if the hash function is not valid, otherwise OK.
+ */
+status_t digest_info_write(const uint8_t *message, const size_t message_len,
```

Comment:
Mark all local function static, otherwise update the naming to follow the naming convention for exportable symbols (~ `rsa_signature_*`).